### PR TITLE
Raghav/get balance

### PR DESCRIPTION
**Pull request by rsundar01**
Implement get_balance, get_task, get_all_tasks

### CODE REVIEW

File: backend/src/main.rs [hunk 1]
```diff
@@ -247,26 +248,98 @@ fn submit_task(
...
+                task_type: task_internal.task_type,
+                payload: task_internal.payload,
+                deadline: task_internal.deadline,
+                reward: task_internal.reward,
+                answers,
+                status: task_internal.status,
+            };
+        }
+
+    });
```
Comment (dsarlis) [inline backend/src/main.rs]: We can simply return `task` from `State.with()` so we can create the task we want in one go (no need to create a dummy one and make mutable so we can change its values later).

File: backend/src/main.rs [hunk 2]
```diff
@@ -247,26 +248,98 @@ fn submit_task(
...
+        let ledger = s.ledger.borrow();
+        if !ledger.contains_key(&caller) {
+            ic_cdk::trap(&format!("{} has not been registered yet.", caller));
+        }
+        let task_map = s.tasks.borrow();
+        if !task_map.contains_key(&id) {
+            ic_cdk::trap(&format!("Requested task id {} cannot be found", id));
+        }
+        let answers_map = s.answers.borrow();
+        if let Some(task_internal_ref) = task_map.get(&id) {
```
Comment (dsarlis) [inline backend/src/main.rs]: Now that I'm reading this again, I think we should do something like:
```
match task_map.get(&id) {
  Some(task_internal_ref) => ...,
  None => ic_cdk::trap(...),
}
```
I can also adjust my PR to do something similar.

File: backend/src/main.rs [hunk 3]
```diff
@@ -247,26 +248,98 @@ fn submit_task(
...
+            let task_internal : TaskInternal = (*task_internal_ref).clone();
+            let mut answers:Vec<Answer> = Vec::new();
+            for ans_id in task_internal.answers {
+                if !answers_map.contains_key(&ans_id) {
+                    ic_cdk::trap(&format!("Inconsistent state. AnswerId {} cannot be found", ans_id));
+                }
+                if let Some(ans_ref) = answers_map.get(&ans_id){
+                    answers.push(ans_ref.clone());
+                }
```
Comment (dsarlis) [inline backend/src/main.rs]: Same here. We can just handle the two outcomes of `answers_map.get(&ans_id)` directly.

**Review by dsarlis — APPROVED** [review]
LGTM. Mostly some nits.

File: backend/src/main.rs (hunk 2 above)
Comment (LaraAS) [reply backend/src/main.rs]: Agree that it might make it a bit more readable

File: backend/src/main.rs [hunk 4]
```diff
@@ -247,26 +248,98 @@ fn submit_task(
...
+            }
+            task = Task {
+                submitter: task_internal.submitter,
+                task_type: task_internal.task_type,
+                payload: task_internal.payload,
+                deadline: task_internal.deadline,
+                reward: task_internal.reward,
+                answers,
```
Comment (LaraAS) [inline backend/src/main.rs]: I am unclear whether the task contains the answers directly or whether it just contains the answer_ids. 
It it the case that the taskInternal only contains the answer_ids but there is also a struct Task which contains the answers directly and we just send this back here?
If this is the case, do we need the distinction at all or couldn't we just use the struct which directly has the answers in the tasks (maybe even without answer_ids)?

File: backend/src/main.rs [hunk 5]
```diff
@@ -247,26 +248,98 @@ fn submit_task(
...
 
 #[query]
 fn get_balance() -> Amount {
-    0
+    let caller = caller();
+    let mut amount : u64 = 0;
+
+    STATE.with(|s| {
+        let ledger = s.ledger.borrow();
+        if !ledger.contains_key(&caller) {
```
Comment (LaraAS) [inline backend/src/main.rs]: I guess here we could again use a match and just trap in the none-case?

**Review by LaraAS — APPROVED** [review]

**Comment by dsarlis** [comment]
There's a problem with `get_task` when I use it from the command line. I get an Invalid IDL error. From a quick debugging session I did, it seems that the blob we return is correct, i.e. if I use it, I can decode the value just fine in the canister and print it. Edge cases, like task id doesn't exist or user is not registered are working fine, so are the other methods, get_balance and get_all_tasks. So, I'll merge this to keep making progress and we can figure out what's the problem.

**Pull request merged**